compiler/rustc_middle/src/lint.rs

```diff
@@ -398,8 +398,16 @@ pub fn lint_level(
             }
         }
 
-        // Finally, run `decorate`.
-        decorate(&mut err);
+        // Finally, run `decorate`. This is guarded by a `can_emit_warnings()` check so that any
+        // `def_path_str` called within `decorate` won't trigger a `must_produce_diag` ICE if the
+        // `err` isn't eventually emitted (e.g. due to `-A warnings`). If an `err` is force-warn,
+        // it's going to be emitted anyway.
+        if matches!(err_level, rustc_errors::Level::ForceWarning(_))
+            || sess.dcx().can_emit_warnings()
+        {
+            decorate(&mut err);
+        }
+
```

Comment:
This isn't entirely correct. The test attached below ICEs after your PR because decorate is incorrectly skipped. `can_emit_warnings` is only relevant when the error level is `Warning`.

```rust
//@ compile-flags: -Dunused_must_use -Awarnings --crate-type=lib
//@ check-pass

#[must_use]
fn f() {}

pub fn g() {
    f();
}
```

Reply:
Also please add a test with `--force-warn` as well.

Reply:
> This isn't entirely correct. The test attached below ICEs after your PR because decorate is incorrectly skipped. `can_emit_warnings` is only relevant when the error level is `Warning`.

Thanks for pointing that out. I forgor to also invoke `decorate` for the actual `Level::Error` case (which lint levels Deny and Forbid correspond to).

Reply:
Added a test case with `--force-warn`.